File: MiniDOMBase/DMD/Anagrafica/CImpieghi.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica
      

    ''' <summary>
    ''' Collezione di impieghi di una persona
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class CImpieghi
        Inherits CCollection(Of CImpiegato)

        <NonSerialized> _
        Private m_Persona As CPersonaFisica

        Public Sub New()
            Me.m_Persona = Nothing
        End Sub

        Public Sub New(ByVal persona As CPersonaFisica)
            Me.New()
            Me.Initialize(persona)
        End Sub

        Public ReadOnly Property Persona As CPersonaFisica
            Get
                Return Me.m_Persona
            End Get
        End Property

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Persona IsNot Nothing) Then DirectCast(value, CImpiegato).SetPersona(Me.m_Persona)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Persona IsNot Nothing) Then DirectCast(newValue, CImpiegato).SetPersona(Me.m_Persona)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Protected Friend Sub SetPersona(ByVal value As CPersonaFisica)
            Me.m_Persona = value
            For Each impiego As CImpiegato In Me
                impiego.Persona = value
            Next
        End Sub

        Public Overloads Function Add() As CImpiegato
            Dim item As New CImpiegato
            item.Stato = ObjectStatus.OBJECT_VALID
            MyBase.Add(item)
            Return item
        End Function

        Public Function AddImpiego(ByVal azienda As CAzienda, ByVal ufficio As String, ByVal posizione As String) As CImpiegato
            Dim item As New CImpiegato
            If (azienda Is Nothing) Then Throw New ArgumentNullException("azienda")
            With item
                .Azienda = azienda
                .Ufficio = ufficio
                .Posizione = posizione
                .Stato = ObjectStatus.OBJECT_VALID
            End With
            MyBase.Add(item)
            Return item
        End Function

        Protected Friend Function Initialize(ByVal persona As CPersonaFisica) As Boolean
            If (persona Is Nothing) Then Throw New ArgumentNullException("persona")

            MyBase.Clear()
            Me.m_Persona = persona

            If GetID(persona) = 0 Then Return True

            Dim conn As CDBConnection = persona.GetConnection
            If (conn.IsRemote) Then
                Dim ret As String = RPC.InvokeMethod("/websvc/modPersone.aspx?_a=getImpieghi", "aid", RPC.int2n(GetID(persona)))
                If (ret <> "") Then Me.AddRange(XML.Utils.Serializer.Deserialize(ret))
            Else
                Dim cursor As CImpiegatiCursor = Nothing
                Try
                    cursor = New CImpiegatiCursor
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                    cursor.PersonaID.Value = Databases.GetID(persona, 0)
                    cursor.ID.SortOrder = SortEnum.SORT_DESC
                    cursor.IgnoreRights = True
                    While Not cursor.EOF
                        MyBase.Add(cursor.Item)
                        cursor.MoveNext()
                    End While
                Catch ex As Exception
                    Sistema.Events.NotifyUnhandledException(ex)
                    Throw
                Finally
                    If (cursor IsNot Nothing) Then cursor.Dispose() : cursor = Nothing
                End Try
            End If


            Return True
        End Function


    End Class






End Class

File: MiniDOMBase/DMD/Anagrafica/Persone/PersoneFisiche/Impieghi/CImpieghi.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;

namespace minidom
{
    public partial class Anagrafica
    {


        /// <summary>
        /// Collezione di impieghi di una persona
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class CImpieghi 
            : CCollection<CImpiegato>
        {
            [NonSerialized] private CPersonaFisica m_Persona;


            /// <summary>
            /// Costruttore
            /// </summary>
            public CImpieghi()
            {
                m_Persona = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="persona"></param>
            public CImpieghi(CPersonaFisica persona) : this()
            {
                Initialize(persona);
            }

            /// <summary>
            /// Restituisce un riferimento alla persona 
            /// </summary>
            public CPersonaFisica Persona
            {
                get
                {
                    return m_Persona;
                }
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, CImpiegato value)
            {
                if (m_Persona is object)
                    value.SetPersona(m_Persona);

                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, CImpiegato oldValue, CImpiegato newValue)
            {
                if (m_Persona is object)
                    newValue.SetPersona(m_Persona);

                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Imposta la persona a cui appartiene la collezione
            /// </summary>
            /// <param name="value"></param>
            protected internal void SetPersona(CPersonaFisica value)
            {
                m_Persona = value;
                foreach (CImpiegato impiego in this)
                    impiego.Persona = value;
            }

            /// <summary>
            /// Aggiunge un nuovo impiego vuoto
            /// </summary>
            /// <returns></returns>
            public CImpiegato Add()
            {
                var item = new CImpiegato();
                item.Stato = ObjectStatus.OBJECT_VALID;
                Add(item);
                return item;
            }

            /// <summary>
            /// Aggiunge un nuovo impiego
            /// </summary>
            /// <param name="azienda"></param>
            /// <param name="ufficio"></param>
            /// <param name="posizione"></param>
            /// <returns></returns>
            public CImpiegato AddImpiego(CAzienda azienda, string ufficio, string posizione)
            {
                var item = new CImpiegato();
                if (azienda is null)
                    throw new ArgumentNullException("azienda");
                item.Azienda = azienda;
                item.Ufficio = ufficio;
                item.Posizione = posizione;
                item.Stato = ObjectStatus.OBJECT_VALID;
                Add(item);
                return item;
            }

            /// <summary>
            /// Carica gli impieghi
            /// </summary>
            /// <param name="persona"></param>
            /// <returns></returns>
            protected internal bool Initialize(CPersonaFisica persona)
            {
                if (persona is null)
                    throw new ArgumentNullException("persona");
                Clear();
                m_Persona = persona;
                if (DBUtils.GetID(persona, 0) == 0)
                    return true;
                using (var cursor = new CImpiegatiCursor())
                { 
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.PersonaID.Value = DBUtils.GetID(persona, 0);
                    cursor.ID.SortOrder = SortEnum.SORT_DESC;
                    cursor.IgnoreRights = true;
                    while (cursor.Read())
                    {
                        Add(cursor.Item);
                    }
                }
                     
                
                return true;
            }
        }
    }
}